### PR TITLE
BCDA 3048: Since parameter validation

### PR DESCRIPTION
**Pull request by PatKelsh**
### Fixes [BCDA-3048](https://jira.cms.gov/browse/BCDA-3048)

Changed the validation for since parameter to reject malformed or inputs

### Proposed Changes

* Changed since validation to use regex provided by the FHIR specification
* Added unit tests to cover edge cases of since parameter

### Security Implications

No PII/PHI implications

- [ ] new software dependencies

- [ ] security controls or supporting software altered

- [ ] new data stored or transmitted

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

- [x] no PHI/PII is affected by this change

### Acceptance Validation

Tested locally and got a appropriate acceptance or rejection for different `_since` parameters

### CODE REVIEW

File: bcda/web/api.go [hunk 1]
```diff
@@ -300,8 +303,9 @@ func validateRequest(r *http.Request) ([]string, *fhirmodels.OperationOutcome) {
 	// validate optional "_since" parameter
 	params, ok = r.URL.Query()["_since"]
 	if ok {
-		_, err := fhirutils.ParseDate(params[0])
-		if err != nil {
+		instantRegex := `([0-9]([0-9]([0-9][1-9]|[1-9]0)|[1-9]00)|[1-9]000)-(0[1-9]|1[0-2])-(0[1-9]|[1-2][0-9]|3[0-1])T([01][0-9]|2[0-3]):[0-5][0-9]:([0-5][0-9]|60)(\.[0-9]+)?(Z|(\+|-)((0[0-9]|1[0-3]):[0-5][0-9]|14:00))`
+		validSince, err := regexp.MatchString(instantRegex, params[0])
+		if err != nil || !validSince {
 			oo := responseutils.CreateOpOutcome(responseutils.Error, responseutils.Exception, responseutils.FormatErr, "Invalid date format supplied in _since parameter.  Date must be in FHIR DateTime format.")
```
Comment (msnook) [inline bcda/web/api.go:309]: Error response indicates "Date must be in FHIR DateTime format."
Should this be updated to specify "Date must be in FHIR Instant format."?

File: bcda/web/api.go [hunk 2]
```diff
@@ -4,18 +4,21 @@ import (
...
 	"net/http"
 	"os"
 	"strings"
 	"time"
 
 	"github.com/bgentry/que-go"
 	fhirmodels "github.com/eug48/fhir/models"
-	fhirutils "github.com/eug48/fhir/utils"
+
+	// fhirutils "github.com/eug48/fhir/utils"
```
Comment (msnook) [inline bcda/web/api.go:21]: Can this be removed completely, instead of being commented out?

File: bcda/web/api.go [hunk 3]
```diff
@@ -300,8 +303,9 @@ func validateRequest(r *http.Request) ([]string, *fhirmodels.OperationOutcome) {
 	// validate optional "_since" parameter
 	params, ok = r.URL.Query()["_since"]
 	if ok {
-		_, err := fhirutils.ParseDate(params[0])
-		if err != nil {
+		instantRegex := `([0-9]([0-9]([0-9][1-9]|[1-9]0)|[1-9]00)|[1-9]000)-(0[1-9]|1[0-2])-(0[1-9]|[1-2][0-9]|3[0-1])T([01][0-9]|2[0-3]):[0-5][0-9]:([0-5][0-9]|60)(\.[0-9]+)?(Z|(\+|-)((0[0-9]|1[0-3]):[0-5][0-9]|14:00))`
```
Comment (msnook) [inline bcda/web/api.go:306]: Does it make sense to have a comment here with a link to the bulk fhir spec showing that we didn't just "make up" this regex?
https://www.hl7.org/fhir/datatypes.html#primitive

File: bcda/web/api_test.go [hunk 4]
```diff
@@ -190,6 +190,72 @@ func (s *APITestSuite) TestBulkCoverageRequestInvalidSinceFormat() {
...
+	s.TearDownTest()
+	s.SetupTest()
+	bulkCoverageRequestInvalidSinceFormatHelper("Group/all", since, s)
+	s.TearDownTest()
+	s.SetupTest()
+	bulkCoverageRequestInvalidSinceFormatHelper("Group/new", since, s)
+}
+
+func (s *APITestSuite) TestBulkCoverageRequestInvalidSinceFormatOnlyInvalidDate() {
+	since := "2020-04-0"
```
Comment (msnook) [inline bcda/web/api_test.go:249]: 😄

**Comment by msnook** [comment]
Thanks for addressing the comments.  Does it make sense to deploy to `dev` and see how it functions in AWS?
